src/scheduler/pbs_sched.cpp

```diff
@@ -530,204 +506,139 @@ close_server(sched_svrconn *sconn)
-		goto rerr;
-	pbs_errno = 0;
-	reply = PBSD_rdrpy(sock);
-	if (reply == NULL)
-		goto rerr;
-	if (pbs_errno != 0) {
-		char *errmsg = get_conn_errtxt(sock);
-		if (errmsg) {
-			log_eventf(PBSEVENT_SYSTEM | PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_NOTICE,
-				   msg_daemonname, "Server rejected register request for %s connection with error: %s",
-				   is_primary ? "primary" : "secondary", errmsg);
-			die(-1);
-		}
-		goto rerr;
+	/* free qrun_list */
+	for (i = 0; i < qrun_list_size; i++) {
+		if (qrun_list[i].jid != NULL)
+			free(qrun_list[i].jid);
 	}
+	free(qrun_list);
 
-	PBSD_FreeReply(reply);
-	return sock;
+	sched_svr_init();
 
-rerr:
-	pbs_disconnect(sock);
-	PBSD_FreeReply(reply);
-	return -1;
+	clust_primary_sock = -1;
+	clust_secondary_sock = -1;
 }
 
 /**
- * @brief connect to given server, create and add server struct in servers list
- *        also add secondary connection to poll list
+ * @brief connect to all the servers configured. Also add secondary connections
+ *              of individual servers secondary sd's to the poll list
  *
- * @param[in] svrhost - address of server in host[:port] format
  *
  * @return void
  */
 static void
-connect_server(const char *svrhost)
+connect_svrpool()
 {
-	int primary_sock = -1;
-	int secondary_sock = -1;
-	sched_svrconn *svr;
-	void *svrs_newarr;
-
-	if (svrhost == NULL)
-		return;
-
-	while (primary_sock < 0 || secondary_sock < 0) {
-		if (primary_sock < 0) {
-			primary_sock = connect_server_helper(svrhost, 1);
-			if (primary_sock < 0) {
+	int i;
+	svr_conn_t *svr_conns_primary = NULL;
+	svr_conn_t *svr_conns_secondary = NULL;
+	int registered = 0;
+	int num_conf_svrs = get_num_servers();
+
+
+	while (clust_primary_sock < 0 || clust_secondary_sock < 0 || registered == 0) {
+		/* pbs_connect() will return a connection handle for all servers */
+		if (clust_primary_sock < 0) {
+			clust_primary_sock = pbs_connect(NULL);
+			if (clust_primary_sock < 0) {
 				/* wait for 2s for not to burn too much CPU, and then retry connection */
 				sleep(2);
 				continue;
 			}
 		}
-		secondary_sock = connect_server_helper(svrhost, 0);
-		if (secondary_sock < 0) {
+		clust_secondary_sock = pbs_connect(NULL);
+		if (clust_secondary_sock < 0) {
 			/* wait for 2s for not to burn too much CPU, and then retry connection */
 			sleep(2);
 			continue;
 		}
-	}
-	if (tpp_em_add_fd(poll_context, secondary_sock, EM_IN | EM_HUP | EM_ERR) < 0) {
-		log_errf(errno, __func__, "Couldn't add secondary connection to poll list for server %s", svrhost);
-		pbs_disconnect(primary_sock);
-		pbs_disconnect(secondary_sock);
-		die(-1);
-	}
 
-	svr = static_cast<sched_svrconn *>(malloc(sizeof(sched_svrconn)));
-	if (!svr) {
-		log_err(errno, __func__, MEM_ERR_MSG);
-		die(-1);
-	}
+		svr_conns_primary =  static_cast<svr_conn_t *>(get_conn_svr_instances(clust_primary_sock));
+		svr_conns_secondary =  static_cast<svr_conn_t *>(get_conn_svr_instances(clust_secondary_sock));
```

Comment:
Will declaring svr_conns_primary and svr_conns_secondary as 'auto' variables help remove the static_casts here?

Reply:
All the scheduler code is using static_cast in many places. I followed the same approach. Don't know the exact reason. May be we prefer explicit and well defined casting. It is better if we follow one style so that is consistent across all scheduler code and hence leaving this as is.